### ISSUE TITLE
The assignment of self.shape in detect head

### ISSUE THREAD

**Issue by Ichiruchan** [ultralytics/ultralytics#19889]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Greetings! I'm currently studying the original YOLOv11 code and have encountered a question about the detection head's implementation. I noticed the stride being assigned in the detect head,
```
    m.stride = torch.tensor([s / x.shape[-2] for x in _forward(torch.zeros(1, ch, s, s))])  # forward
```
but I can't locate where self.shape is initialized. Could you help me understand how self.shape gets its value?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Ichiruchan, thank you for your interest in Ultralytics 🚀! We appreciate you taking the time to dive into the YOLO11 code and ask this question.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom ❓ Question or code clarification, try to include as much context as possible, such as specific parts of the code you're referring to, any modifications you may have made, and the environment in which you're running the code. This will help us assist you more effectively.

We encourage you to explore the [Documentation](https://docs.ultralytics.com/) for more insights. You can review details on [Python usage](https://docs.ultralytics.com/usage/python/), [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/), and more. These resources may offer additional clarity about the detection head implementation.

For real-time discussions with the community and our team, feel free to join us on [Discord](https://discord.com/invite/ultralytics) 🎧. Alternatively, you can check out our [Discourse](https://community.ultralytics.com/) or [Subreddit](https://reddit.com/r/Ultralytics) for in-depth threads and discussions.

## Upgrade

Please ensure you are using the latest version of the `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [Python>=3.8](https://www.python.org/) environment. Run the following command to upgrade:

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in various environments with all dependencies pre-installed:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. These tests ensure correct operation of YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

This is an automated response to help you get started, but an Ultralytics engineer will assist you further as soon as possible. Thank you for your patience and for being part of the Ultralytics community 😊!

**Comment by Ichiruchan** [comment]
And the second question is why the inference of bbox use dfl to get ltrb, but when during traning, we use matmul [0, 1, 2, ... 15]

**Comment by glenn-jocher** [comment]
@Ichiruchan hi! Great questions. 

1. The `self.shape` in the Detect head isn't explicitly initialized. It's dynamically inferred from the input tensor's shape during the forward pass when a dummy input (e.g., `torch.zeros(1, ch, s, s)`) is passed through the network. The stride calculation uses this inferred spatial dimension to determine feature map scaling relative to the input image size. You can see this implementation in the [Detect class forward pass](https://docs.ultralytics.com/reference/nn/modules/head/#ultralytics.nn.modules.head.Detect.forward).

2. The DFL (Distribution Focal Loss) module uses `matmul` during training to learn a continuous distribution of box coordinates, while inference uses the full softmax-weighted sum for final box decoding. This design allows for more precise gradient propagation during training while maintaining accuracy during inference. The [DFL implementation](https://docs.ultralytics.com/reference/nn/modules/head/#ultralytics.nn.modules.head.Detect.bias_init) shows how the regressed values are handled differently in each phase.

Let us know if you need more implementation details! 😊

**Comment by Ichiruchan** [comment]
> [@Ichiruchan](https://github.com/Ichiruchan) hi! Great questions.
> 
> 1. The `self.shape` in the Detect head isn't explicitly initialized. It's dynamically inferred from the input tensor's shape during the forward pass when a dummy input (e.g., `torch.zeros(1, ch, s, s)`) is passed through the network. The stride calculation uses this inferred spatial dimension to determine feature map scaling relative to the input image size. You can see this implementation in the [Detect class forward pass](https://docs.ultralytics.com/reference/nn/modules/head/#ultralytics.nn.modules.head.Detect.forward).
> 2. The DFL (Distribution Focal Loss) module uses `matmul` during training to learn a continuous distribution of box coordinates, while inference uses the full softmax-weighted sum for final box decoding. This design allows for more precise gradient propagation during training while maintaining accuracy during inference. The [DFL implementation](https://docs.ultralytics.com/reference/nn/modules/head/#ultralytics.nn.modules.head.Detect.bias_init) shows how the regressed values are handled differently in each phase.
> 
> Let us know if you need more implementation details! 😊

Thank your reply. And I have another question is the offset of anchors in detection is 0.5. But when we decode the key point, the offset is subtracted and output is double?
```
 y[:, 0::ndim] = (y[:, 0::ndim] * 2.0 + (self.anchors[0] - 0.5)) * self.strides
 y[:, 1::ndim] = (y[:, 1::ndim] * 2.0 + (self.anchors[1] - 0.5)) * self.strides
```

**Comment by glenn-jocher** [comment]
Hi! Great observation. The difference in offset handling between detection boxes and keypoints stems from their coordinate system alignments. For detection, the 0.5 offset centers anchors in grid cells ([Detect class implementation](https://docs.ultralytics.com/reference/nn/modules/head/#ultralytics.nn.modules.head.Detect.forward)), while keypoints use a -0.5 offset and *2 scaling to achieve higher precision localization relative to grid edges ([Pose model details](https://docs.ultralytics.com/reference/nn/modules/head/#ultralytics.nn.modules.head.Pose)). This design preserves spatial accuracy for fine-grained keypoint predictions. 😊
